src/inmanta/server/services/orchestrationservice.py

```diff
@@ -59,6 +62,133 @@
+        self.old_resource_set = old_resource_set
+
+    def same_resource(self):
+        new_resource = set(self.new_resource).difference("id")
+        old_resource = set(self.old_resource).difference("id")
+        return new_resource == old_resource and all(self.new_resource[k] == self.old_resource[k] for k in new_resource)
+
+
+class PartialUpdateMerger(object):
+    def __init__(
+        self,
+        partial_updates: List[Dict[str, Any]],
+        resource_sets: Dict[ResourceIdStr, Optional[str]],
+        removed_resource_sets: List[str],
+        env: data.Environment,
+    ) -> None:
+        self.partial_updates = partial_updates
+        self.resource_sets = resource_sets
+        self.removed_resource_sets = removed_resource_sets
+        self.env = env
+
+    def _pair_resources_partial_update_to_old_version(
+        self, old_resources: Dict[str, ResourceWithResourceSet], partial_updates: List[Dict[str, Any]]
+    ) -> List[PairedResource]:
+        """
+        returns a list of paired resources
+
+        :param old_resources: a list of tuples containing the resource in the previous version of the model
+        and its resource_set.
+        :param partial_updates: The list of resources part of the partial compile.
+        """
+        paired_resources: List[PairedResource] = []
+        for partial_update in partial_updates:
+            key = Id.parse_id(partial_update["id"]).resource_str()
+            resource_set = self.resource_sets.get(key)
+            pair: PairedResource = PairedResource(
+                partial_update,
+                None,
+                resource_set,
+                None,
+            )
+            if key in old_resources:
+                pair.old_resource = old_resources[key].resource
+                pair.old_resource_set = old_resources[key].resource_set
+            paired_resources.append(pair)
+        return paired_resources
+
+    async def _get_old_resources(self) -> Dict[str, ResourceWithResourceSet]:
+        old_data = await data.Resource.get_resources_in_latest_version(environment=self.env.id)
+        result: Dict[str, ResourceWithResourceSet] = {}
+        for res in old_data:
+            resource: Dict[str, ResourceVersionIdStr] = {
+                "id": res.resource_version_id,
+            }
+            resource.update(res.attributes)
+            result[res.resource_id] = ResourceWithResourceSet(resource, res.resource_set)
+        return result
+
+    async def merge_partial_with_old(self) -> List[Any]:
+
+        old_resources: Dict[str, ResourceWithResourceSet] = await self._get_old_resources()
+        paired_resources: List[PairedResource] = self._pair_resources_partial_update_to_old_version(
+            old_resources, self.partial_updates
+        )
+        updated_resource_sets: Set[str] = set(res.new_resource_set for res in paired_resources if res.new_resource_set)
+
+        def copy_with_incremented_version(resource: Dict[str, Any]) -> Dict[str, Any]:
```

Comment:
This implementation builds on top of the assumption that the version associated with each resource in the resources argument of the put_partial endpoint, matches the version argument of the put_partial endpoint. But I don't think we currently have a check that validates whether that assumption is satisfied. Do you add such a check as well?